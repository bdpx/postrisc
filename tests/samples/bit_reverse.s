.text
    write	"test bit reverse instruction (permb)"
    alloc	80
    ldi.l	%r55, 0x1234567890ABCDEF
    write	"initial value: %x64(r55)"
    permb	%r55, %r55, 63
    permb	%r56, %r78, 63
    write	"r55 %x64(r55) %b64(r55)"
    permb	%r55, %r55, 63
    write	"r55 %x64(r55) %b64(r55)"

    permb	%r56, %r55, 0b111111 ;63
    write	"reverse bits: %x64(r56)"

    permb	%r56, %r55, 0b111110  ;32+16+8+4+2
    write	"reverse bit-pairs: %x64(r56)"

    permb	%r56, %r55, 0b111100  ;32+16+8+4
    write	"reverse nibbles (4-bits): %x64(r56)"

    permb	%r56, %r55, 0b111000 ;32+16+8
    write	"reverse 1bytes: %x64(r55) => %x64(r56)"

    permb	%r56, %r55, 0b110000  ;32+16
    write	"reverse 2bytes: %x64(r55) => %x64(r56)"

    permb	%r56, %r55, 0b100000  ;32
    write	"reverse 4bytes: %x64(r55) => %x64(r56)"
.end
