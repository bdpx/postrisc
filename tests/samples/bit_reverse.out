test bit reverse instruction (permb)
initial value: 1234567890abcdef
r55 f7b3d5091e6a2c48 11110111_10110011_11010101_00001001_00011110_01101010_00101100_01001000
r55 1234567890abcdef 00010010_00110100_01010110_01111000_10010000_10101011_11001101_11101111
reverse bits: f7b3d5091e6a2c48
reverse bit-pairs: fb73ea062d951c84
reverse nibbles (4-bits): fedcba0987654321
reverse 1bytes: 1234567890abcdef => efcdab9078563412
reverse 2bytes: 1234567890abcdef => cdef90ab56781234
reverse 4bytes: 1234567890abcdef => 90abcdef12345678
